File: leetcode/0621-task-scheduler/solution.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        from collections import Counter
        import heapq

        freq = Counter(tasks)
        
        heap = []
        for task in freq:
            heapq.heappush(heap, (-freq[task], task))

        res = 0

        while heap:
            used = 0
            tmp= []
            for i in range(n+1):
                if not heap:
                    break
                
                count,task = heapq.heappop(heap)
                count = -count
                count -=1
                used+=1
                if count > 0:
                    tmp.append((-count, task))
                    

            # res += subres
            for item in tmp:
                heapq.heappush(heap,item)
                tmp = []
            if heap:
                res += n + 1
            else:
                res += used
        return res
        
        

File: leetcode/0621-task-scheduler/solution.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char, int> freq;
        priority_queue<int> pq;

        for (char t: tasks) {
            freq[t]++;
        }
        
        for(const auto &[c, cnt]: freq) {
            pq.push(cnt);
        }

        int res = 0;
        while (!pq.empty()) {
            int used = 0;

            vector<int> temp;
            for(int i = 0; i < n + 1; i++) {
                if (pq.empty()) {
                    break;
                }
                else {
                    int x = pq.top();
                    pq.pop();  
                    x--;
                    used++;
                    if (x > 0) {
                        temp.push_back(x);
                    }                         
                }
            }
            for (int t: temp) {
                pq.push(t);
            }
            if (pq.empty()) {
                res += used;
            }
            else {
                res += n + 1;
            }

        }
        return res;
    }
};

